### PR TITLE
`GeoSeries.contains` that matches `GeoPandas` with the exception of cases that depend on intersection.

### PR DESCRIPTION
**Pull request by thomcom**
## Description
Closes https://github.com/rapidsai/cuspatial/issues/738, https://github.com/rapidsai/cuspatial/issues/742, https://github.com/rapidsai/cuspatial/issues/743, and https://github.com/rapidsai/cuspatial/issues/744>

A feature $a$ is contained in a feature $b$ $\leftrightarrow$ all of the points in $a$ are in the inner region of $b$ and none of the exterior points of $a$ or $b$ intersect. This PR implements the first part of the `.contains` operation, that is, $\leftrightarrow$ all of the points in $a$ are in the inner region of $b$.

We do not have an intersection primitive available yet, so I will add the second portion of this operation when it is available. There are a small number of pytest `xfails` that demonstrate where we can't agree with GeoPandas yet.

This PR adds boundary exclusion to our point-in-polygon algorithm. Previously, point-in-polygon had inconsistent inclusion or exclusion of the polygon boundary, based on floating-point error and implementation details of the RayCrossings algorithm. Due to that inconsistency I've added explicit boundary exclusion.

Boundary exclusion in this context means that any point in $b$ that is colinear with a line segment in $a$ is treated as an intersection, which then violates the requirements of `.contains`, returning false for $a.contains(b)$.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cuspatial/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by harrism** [comment]
In the description, can you explain what "with the exception of exterior cases" means, precisely?

File: cpp/include/cuspatial/experimental/detail/point_in_polygon.cuh [hunk 1]
```diff
@@ -72,22 +73,36 @@ __device__ inline bool is_point_in_polygon(Cart2d const& test_point,
...
-      if (y1_flag != y0_flag) {
-        T run           = b.x - a.x;
-        T rise          = b.y - a.y;
-        T rise_to_point = test_point.y - a.y;
+      T run          = b.x - a.x;
+      T rise         = b.y - a.y;
+
+      // Points on the line segment are the same, so intersection is impossible.
+      // This is possible because we allow closed or unclosed polygons.
+      if (run == 0.0 && rise == 0.0) continue;
```
Comment (isVoid) [inline cpp/include/cuspatial/experimental/detail/point_in_polygon.cuh:81]: Can you try to replace this compare with the new utility I added in https://github.com/rapidsai/cuspatial/pull/773?
I'd use as `if (float_eq_by_ulp(b.x, a.x) && float_eq_by_ulp(b.y, a.y)) continue;`

Comment (isVoid) [reply cpp/include/cuspatial/experimental/detail/point_in_polygon.cuh:81]: (No need to push to this PR for now - just want to test how usable that utility is)

Comment (thomcom) [reply cpp/include/cuspatial/experimental/detail/point_in_polygon.cuh:81]: I was able to test it. Looks like it improves our accuracy for detecting those really hard corner cases like @harrism had identified.

```
            [Point([3.33, 1.11]), Polygon([[6, 2], [3, 1], [3, 4], [6, 2]]), True],
        ],
    )
    def test_float_precision_limits(point, polygon, expects):
        gpdpoint = gpd.GeoSeries(point)
        gpdpolygon = gpd.GeoSeries(polygon)
        point = cuspatial.from_geopandas(gpdpoint)
        polygon = cuspatial.from_geopandas(gpdpolygon)
        result = polygon.contains(point)
>       assert gpdpolygon.contains(gpdpoint).values == result.values_host
```

Produces a false now, previously True. It is good for us to have improved accuracy but in this case we conflict with Pandas, who produces True.

Comment (harrism) [reply cpp/include/cuspatial/experimental/detail/point_in_polygon.cuh:81]: Isn't this because GeoPandas `contains` uses Shapely `contains` and not `contains_properly`? By excluding boundary points, you are now implementing `contains_properly`, not `contains`. https://shapely.readthedocs.io/en/latest/reference/shapely.contains_properly.html

**Comment by thomcom** [comment]
Closing in favor of #749
